### PR TITLE
Fanling/cadl dpg anomalydetector

### PR DESCRIPTION
**Pull request by happyling123**
# Description

Please add an informative description that covers that changes made by the pull request and link all relevant issues.

If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.

# All SDK Contribution checklist:
- [ ] **The pull request does not introduce [breaking changes]**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

**Review by minkefusiji — APPROVED** [review]

File: sdk/anomalydetector/azure-ai-anomalydetector/src/samples/java/com/azure/ai/anomalydetector/ReadmeSamples.java [hunk 1]
```diff
@@ -3,9 +3,43 @@
...
+        AddHeadersPolicy addHeadersPolicy = new AddHeadersPolicy(headers);
+
+        HttpPipeline httpPipeline = new HttpPipelineBuilder().httpClient(HttpClient.createDefault())
+            .policies(authPolicy, addHeadersPolicy).build();
+        // Instantiate a client that will be used to call the service.
+        HttpLogOptions httpLogOptions = new HttpLogOptions();
+        httpLogOptions.setLogLevel(HttpLogDetailLevel.BODY_AND_HEADERS);
+
+        AnomalyDetectorClient anomalyDetectorClient = new AnomalyDetectorClientBuilder()
+            .pipeline(httpPipeline)
```
Comment (weidongxu-microsoft) [inline sdk/anomalydetector/azure-ai-anomalydetector/src/samples/java/com/azure/ai/anomalydetector/ReadmeSamples.java:39]: any reason to choose use the HttpPipeline instead of using "credential()" method in builder?

File: sdk/anomalydetector/azure-ai-anomalydetector/src/samples/java/com/azure/ai/anomalydetector/DetectAnomaliesLastPoint.java [hunk 2]
```diff
@@ -56,9 +57,10 @@ public static void main(final String[] args) throws IOException {
         InputStream fileInputStream = new FileInputStream("azure-ai-anomalydetector\\src\\samples\\java\\sample_data\\request-data.json");
         JsonReader reader = Json.createReader(fileInputStream);
         BinaryData detectBody = BinaryData.fromString(reader.readObject().toString());
-        RequestOptions requestOptions = new RequestOptions();
-        Response<BinaryData> response = anomalyDetectorClient.detectUnivariateLastPointWithResponse(detectBody, requestOptions);
-        System.out.println(response.getValue().toString());
 
+        DetectRequest detectRequest = detectBody.toObject(DetectRequest.class);
+        LastDetectResponse lastDetectResponse = anomalyDetectorClient.detectUnivariateLastPoint(detectRequest);
```
Comment (weidongxu-microsoft) [inline sdk/anomalydetector/azure-ai-anomalydetector/src/samples/java/com/azure/ai/anomalydetector/DetectAnomaliesLastPoint.java:62]: If your sample recommend user to load the JSON, then call the API, why do you need model in the first place...

File: sdk/anomalydetector/azure-ai-anomalydetector/src/samples/java/com/azure/ai/anomalydetector/ReadmeSamples.java [hunk 3]
```diff
@@ -3,9 +3,43 @@
...
-        // BEGIN: com.azure.ai.anomalydetector.readme
-        // END: com.azure.ai.anomalydetector.readme
+
+    private static void createClient() {
+        // BEGIN: readme-sample-createAnomalyDetectorClient
+        String endpoint = "<anomaly-detector-resource-endpoint>";
+        String key = "<anomaly-detector-resource-key>";
+
+        HttpHeaders headers = new HttpHeaders()
+            .put("Accept", ContentType.APPLICATION_JSON);
```
Comment (weidongxu-microsoft) [inline sdk/anomalydetector/azure-ai-anomalydetector/src/samples/java/com/azure/ai/anomalydetector/ReadmeSamples.java:26]: Is adding the "accept" header necessary?

File: sdk/anomalydetector/azure-ai-anomalydetector/src/samples/java/com/azure/ai/anomalydetector/DetectAnomaliesLastPoint.java (hunk 2 above)
Comment (happyling123) [reply sdk/anomalydetector/azure-ai-anomalydetector/src/samples/java/com/azure/ai/anomalydetector/DetectAnomaliesLastPoint.java:62]: Yes. That is a question. In fact, I think Json is better to work with, in comparison with setting object attributes one by one (I tried but gave up.).

File: sdk/anomalydetector/azure-ai-anomalydetector/src/samples/java/com/azure/ai/anomalydetector/ReadmeSamples.java (hunk 1 above)
Comment (happyling123) [reply sdk/anomalydetector/azure-ai-anomalydetector/src/samples/java/com/azure/ai/anomalydetector/ReadmeSamples.java:39]: follow the previous version.

File: sdk/anomalydetector/azure-ai-anomalydetector/src/samples/java/com/azure/ai/anomalydetector/ReadmeSamples.java (hunk 3 above)
Comment (happyling123) [reply sdk/anomalydetector/azure-ai-anomalydetector/src/samples/java/com/azure/ai/anomalydetector/ReadmeSamples.java:26]: follow the previous version.

File: sdk/anomalydetector/azure-ai-anomalydetector/src/samples/java/com/azure/ai/anomalydetector/ReadmeSamples.java (hunk 1 above)
Comment (weidongxu-microsoft) [reply sdk/anomalydetector/azure-ai-anomalydetector/src/samples/java/com/azure/ai/anomalydetector/ReadmeSamples.java:39]: This is the sample, so "follow the previous version." is not an answer. You need to know why.

File: sdk/anomalydetector/azure-ai-anomalydetector/src/samples/java/com/azure/ai/anomalydetector/DetectAnomaliesLastPoint.java (hunk 2 above)
Comment (happyling123) [reply sdk/anomalydetector/azure-ai-anomalydetector/src/samples/java/com/azure/ai/anomalydetector/DetectAnomaliesLastPoint.java]: Update all.

File: sdk/anomalydetector/azure-ai-anomalydetector/src/samples/java/com/azure/ai/anomalydetector/ReadmeSamples.java (hunk 1 above)
Comment (minkefusiji) [reply sdk/anomalydetector/azure-ai-anomalydetector/src/samples/java/com/azure/ai/anomalydetector/ReadmeSamples.java:39]: Hi @weidongxu-microsoft , why should we choose "credential" here?
Actually, we use ”AzureKeyCredential“ in HttpPipeline.

        _**HttpPipelinePolicy authPolicy = new AzureKeyCredentialPolicy("Ocp-Apim-Subscription-Key",
            new AzureKeyCredential(key));
        AddHeadersPolicy addHeadersPolicy = new AddHeadersPolicy(headers);
        HttpPipeline httpPipeline = new HttpPipelineBuilder().httpClient(HttpClient.createDefault())
            .policies(authPolicy, addHeadersPolicy).build();**_

File: sdk/anomalydetector/azure-ai-anomalydetector/README.md [hunk 4]
```diff
@@ -59,6 +59,30 @@ You will also need to [register a new AAD application][register_aad_app] and [gr
...
+AddHeadersPolicy addHeadersPolicy = new AddHeadersPolicy(headers);
+
+HttpPipeline httpPipeline = new HttpPipelineBuilder().httpClient(HttpClient.createDefault())
+    .policies(authPolicy, addHeadersPolicy).build();
+// Instantiate a client that will be used to call the service.
+HttpLogOptions httpLogOptions = new HttpLogOptions();
+httpLogOptions.setLogLevel(HttpLogDetailLevel.BODY_AND_HEADERS);
+
+AnomalyDetectorClient anomalyDetectorClient = new AnomalyDetectorClientBuilder()
+    .pipeline(httpPipeline)
```
Comment (srnagar) [inline sdk/anomalydetector/azure-ai-anomalydetector/README.md:81]: Why do we have a sample showing setting the pipeline? This pipeline is configured with just a couple of policies and missing many important ones. We should not include this sample in README.

File: sdk/anomalydetector/azure-ai-anomalydetector/src/samples/java/com/azure/ai/anomalydetector/ReadmeSamples.java (hunk 1 above)
Comment (srnagar) [reply sdk/anomalydetector/azure-ai-anomalydetector/src/samples/java/com/azure/ai/anomalydetector/ReadmeSamples.java:39]: I had the same comment above in README. We should use `credential` here and not use HttpPipeline.

File: sdk/anomalydetector/azure-ai-anomalydetector/src/samples/java/com/azure/ai/anomalydetector/ReadmeSamples.java (hunk 3 above)
Comment (srnagar) [reply sdk/anomalydetector/azure-ai-anomalydetector/src/samples/java/com/azure/ai/anomalydetector/ReadmeSamples.java:26]: Is this required for creating the client?

File: sdk/anomalydetector/azure-ai-anomalydetector/src/samples/java/com/azure/ai/anomalydetector/ReadmeSamples.java (hunk 1 above)
Comment (weidongxu-microsoft) [reply sdk/anomalydetector/azure-ai-anomalydetector/src/samples/java/com/azure/ai/anomalydetector/ReadmeSamples.java:39]: @minkefusiji 

I think I even generated the samples for you about the "credential()"? https://github.com/Azure/azure-sdk-for-java/blob/cadl_dpg_anomalydetector/sdk/anomalydetector/azure-ai-anomalydetector/src/samples/java/com/azure/ai/anomalydetector/generated/GetMultivariateModel.java#L17-L21

Of course they may need changes, but at least take it as what you start?

File: sdk/anomalydetector/azure-ai-anomalydetector/src/samples/java/com/azure/ai/anomalydetector/ReadmeSamples.java (hunk 3 above)
Comment (weidongxu-microsoft) [reply sdk/anomalydetector/azure-ai-anomalydetector/src/samples/java/com/azure/ai/anomalydetector/ReadmeSamples.java:26]: Usually it is already set in code https://github.com/Azure/azure-sdk-for-java/blob/cadl_dpg_anomalydetector/sdk/anomalydetector/azure-ai-anomalydetector/src/main/java/com/azure/ai/anomalydetector/implementation/AnomalyDetectorClientImpl.java#L453

File: sdk/anomalydetector/azure-ai-anomalydetector/src/samples/java/com/azure/ai/anomalydetector/MultivariateSample.java [hunk 5]
```diff
@@ -193,37 +197,17 @@ public static void run(BinaryData trainBody, BinaryData beginInferBody) throws E
         InputStream fileInputStream = new FileInputStream("azure-ai-anomalydetector\\src\\samples\\java\\sample_data\\sync_infer_body.json");
```
Comment (weidongxu-microsoft) [inline sdk/anomalydetector/azure-ai-anomalydetector/src/samples/java/com/azure/ai/anomalydetector/MultivariateSample.java:197]: @srnagar Where do we recommend to store sample data?

Currently it is here azure-ai-anomalydetector/src/samples/java/sample_data

Should it be in azure-ai-anomalydetector/src/samples/resources ?

File: sdk/anomalydetector/azure-ai-anomalydetector/src/samples/java/com/azure/ai/anomalydetector/MultivariateSample.java [hunk 6]
```diff
@@ -85,105 +91,103 @@ private static AnomalyDetectorClient getClient(String endpoint, String key) {
...
+        OffsetDateTime endTime = OffsetDateTime.of(2021, 1, 2, 5, 0, 0, 0, ZoneOffset.UTC);
+        ModelInfo trainRequest = new ModelInfo(datasource, startTime, endTime);
+        trainRequest.setSlidingWindow(200);
+        AlignPolicy alignPolicy = new AlignPolicy();
+        alignPolicy.setAlignMode(AlignMode.OUTER);
+        alignPolicy.setFillNAMethod(FillNAMethod.LINEAR);
+        alignPolicy.setPaddingValue(0.0);
+        trainRequest.setAlignPolicy(alignPolicy);
+        trainRequest.setDataSchema(dataSchema);
+        trainRequest.setDisplayName("SampleRequest");
```
Comment (weidongxu-microsoft) [inline sdk/anomalydetector/azure-ai-anomalydetector/src/samples/java/com/azure/ai/anomalydetector/MultivariateSample.java:153]: You can use the fluent style
e.g.
```java
trainRequest.setSlidingWindow(200)
    .setAlignPolicy(new AlignPolicy()
        .setAlignMode(AlignMode.OUTER)...)....
```

PS: is it correct that all properties in AlignPolicy be optional?

File: sdk/anomalydetector/azure-ai-anomalydetector/src/samples/java/com/azure/ai/anomalydetector/MultivariateSample.java [hunk 7]
```diff
@@ -85,105 +91,103 @@ private static AnomalyDetectorClient getClient(String endpoint, String key) {
...
                 break;
             } else if (modelStatus == ModelStatus.FAILED) {
                 System.out.println("FAILED");
-                throw new Exception(modelInfo.getJsonArray("errors").getString(0));
+                String errorStr = "";
+                for (ErrorResponse errorResponse : modelInfo.getErrors()) {
+                    System.out.println(errorResponse.getCode() + errorResponse.getMessage());
+                    errorStr += ";" + errorResponse.getCode() + errorResponse.getMessage();
+                }
+                throw new Exception(errorStr);
```
Comment (weidongxu-microsoft) [inline sdk/anomalydetector/azure-ai-anomalydetector/src/samples/java/com/azure/ai/anomalydetector/MultivariateSample.java:172]: Should not throw bare `Exception` in Java.

At least use a `RuntimeException`, or proper subclass of it.

File: sdk/anomalydetector/azure-ai-anomalydetector/src/samples/java/com/azure/ai/anomalydetector/MultivariateSample.java [hunk 8]
```diff
@@ -85,105 +91,103 @@ private static AnomalyDetectorClient getClient(String endpoint, String key) {
...
+    public static void run(String datasource, DataSchema dataSchema) throws Exception {
         String endpoint = "<anomaly-detector-resource-endpoint>";
         String key = "<anomaly-detector-resource-key>";
 
         // Get multivariate client
         AnomalyDetectorClient client = getClient(endpoint, key);
 
+        // set training request
+        OffsetDateTime startTime = OffsetDateTime.of(2021, 1, 2, 0, 0, 0, 0, ZoneOffset.UTC);
+        OffsetDateTime endTime = OffsetDateTime.of(2021, 1, 2, 5, 0, 0, 0, ZoneOffset.UTC);
```
Comment (weidongxu-microsoft) [inline sdk/anomalydetector/azure-ai-anomalydetector/src/samples/java/com/azure/ai/anomalydetector/MultivariateSample.java:144]: nit, `parse` might be easier to read https://docs.oracle.com/javase/8/docs/api/java/time/OffsetDateTime.html#parse-java.lang.CharSequence-

This is totally up to you.

File: sdk/anomalydetector/azure-ai-anomalydetector/src/samples/java/com/azure/ai/anomalydetector/MultivariateSample.java [hunk 9]
```diff
@@ -85,105 +91,103 @@ private static AnomalyDetectorClient getClient(String endpoint, String key) {
...
+            PagedResponse<Model> items = ite.next();
             System.out.println("The result in the page " + i);
             i++;
-            for (BinaryData item : items.getValue()) {
-                JsonObject jsonObject = Json.createReader(new StringReader(item.toString())).readObject();
-                System.out.println("\t" + jsonObject.getString("modelId"));
+            for (Model item : items.getValue()) {
+                System.out.println("\t" + item.getModelId());
             }
         }
```
Comment (weidongxu-microsoft) [inline sdk/anomalydetector/azure-ai-anomalydetector/src/samples/java/com/azure/ai/anomalydetector/MultivariateSample.java:127]: See if you can do this via stream API. https://docs.oracle.com/javase/8/docs/api/java/util/stream/Stream.html

File: sdk/anomalydetector/azure-ai-anomalydetector/src/samples/java/com/azure/ai/anomalydetector/ReadmeSamples.java (hunk 1 above)
Comment (happyling123) [reply sdk/anomalydetector/azure-ai-anomalydetector/src/samples/java/com/azure/ai/anomalydetector/ReadmeSamples.java]: fixed it.

File: sdk/anomalydetector/azure-ai-anomalydetector/src/samples/java/com/azure/ai/anomalydetector/MultivariateSample.java (hunk 6 above)
Comment (happyling123) [reply sdk/anomalydetector/azure-ai-anomalydetector/src/samples/java/com/azure/ai/anomalydetector/MultivariateSample.java]: updated.
Yes, AlignPolicy is optional.

File: sdk/anomalydetector/azure-ai-anomalydetector/src/samples/java/com/azure/ai/anomalydetector/MultivariateSample.java (hunk 7 above)
Comment (happyling123) [reply sdk/anomalydetector/azure-ai-anomalydetector/src/samples/java/com/azure/ai/anomalydetector/MultivariateSample.java]: fixed.

File: sdk/anomalydetector/azure-ai-anomalydetector/src/samples/java/com/azure/ai/anomalydetector/MultivariateSample.java (hunk 8 above)
Comment (happyling123) [reply sdk/anomalydetector/azure-ai-anomalydetector/src/samples/java/com/azure/ai/anomalydetector/MultivariateSample.java]: use parse.

File: sdk/anomalydetector/azure-ai-anomalydetector/src/samples/java/com/azure/ai/anomalydetector/MultivariateSample.java (hunk 9 above)
Comment (happyling123) [reply sdk/anomalydetector/azure-ai-anomalydetector/src/samples/java/com/azure/ai/anomalydetector/MultivariateSample.java]: updated.

**Review by weidongxu-microsoft — APPROVED** [review]

File: sdk/anomalydetector/azure-ai-anomalydetector/CHANGELOG.md [hunk 10]
```diff
@@ -1,14 +1,27 @@
...
 
 ### Features Added
 
+- Added `OneTable` and `MultiTable` two data schemas.
+- Added `topContributorCoun` in the request body for `getMultivariateBatchDetectionResult`, `detectMultivariateBatchAnomaly` and `detectMultivariateLastAnomaly`.
+
 ### Breaking Changes
 
-### Bugs Fixed
+- Renamed `trainMultivariateModelWithResponse` to `createAndTrainMultivariateModel`
```
Comment (weidongxu-microsoft) [inline sdk/anomalydetector/azure-ai-anomalydetector/CHANGELOG.md:12]: remember to update to latest name, e.g. currently it is `trainMultivariateModel`

You can do that in another PR.

File: sdk/anomalydetector/azure-ai-anomalydetector/CHANGELOG.md [hunk 11]
```diff
@@ -1,14 +1,27 @@
 # Release History
 
-## 3.0.0-beta.5 (Unreleased)
+## 3.0.0-beta.5 (2022-12-01)
```
Comment (weidongxu-microsoft) [inline sdk/anomalydetector/azure-ai-anomalydetector/CHANGELOG.md:3]: remember to update date

**Review by weidongxu-microsoft — APPROVED** [review]

File: sdk/anomalydetector/azure-ai-anomalydetector/src/test/java/com/azure/ai/anomalydetector/MultivariateAnomalyDetectorClientTest.java [hunk 12]
```diff
@@ -0,0 +1,36 @@
+package com.azure.ai.anomalydetector;
```
Comment (weidongxu-microsoft) [inline sdk/anomalydetector/azure-ai-anomalydetector/src/test/java/com/azure/ai/anomalydetector/MultivariateAnomalyDetectorClientTest.java:1]: why this can pass CI???

Comment (happyling123) [reply sdk/anomalydetector/azure-ai-anomalydetector/src/test/java/com/azure/ai/anomalydetector/MultivariateAnomalyDetectorClientTest.java:1]: I added Copyright. Do you mean that?

File: sdk/anomalydetector/azure-ai-anomalydetector/CHANGELOG.md (hunk 10 above)
Comment (happyling123) [reply sdk/anomalydetector/azure-ai-anomalydetector/CHANGELOG.md:12]: updated.

Comment (weidongxu-microsoft) [reply sdk/anomalydetector/azure-ai-anomalydetector/CHANGELOG.md:12]: The PR is already merged. Please create another PR target main branch.
